File: mail/protocols/imap/src/main/java/com/fsck/k9/mail/store/imap/NegativeImapResponseException.java
package com.fsck.k9.mail.store.imap;

import java.util.List;

import com.fsck.k9.mail.MessagingException;

import static com.fsck.k9.mail.store.imap.ImapResponseParser.equalsIgnoreCase;


class NegativeImapResponseException extends MessagingException {
    private static final long serialVersionUID = 3725007182205882394L;

    private final List<ImapResponse> responses;
    private String alertText;

    public NegativeImapResponseException(String message, List<ImapResponse> responses) {
        super(message, true);
        this.responses = responses;
    }

    public String getAlertText() {
        if (alertText == null) {
            ImapResponse lastResponse = getLastResponse();
            alertText = AlertResponse.getAlertText(lastResponse);
        }
        
        return alertText;
    }

    public boolean wasByeResponseReceived() {
        for (ImapResponse response : responses) {
            if (response.getTag() == null && response.size() >= 1 && equalsIgnoreCase(response.get(0), Responses.BYE)) {
                return true;
            }
        }

        return false;
    }

    public ImapResponse getLastResponse() {
        return responses.get(responses.size() - 1);
    }
}

File: mail/protocols/imap/src/main/java/com/fsck/k9/mail/store/imap/NegativeImapResponseException.kt
package com.fsck.k9.mail.store.imap

import com.fsck.k9.mail.MessagingException

internal class NegativeImapResponseException(
    message: String?,
    private val responses: List<ImapResponse>,
) : MessagingException(message, true) {

    init {
        require(responses.isNotEmpty()) { "List of responses must not be empty" }
    }

    val lastResponse: ImapResponse
        get() = responses.last()

    val responseText: String? by lazy { ResponseTextExtractor.getResponseText(lastResponse) }

    val alertText: String? by lazy { AlertResponse.getAlertText(lastResponse) }

    fun wasByeResponseReceived(): Boolean {
        return responses.any { it.isByeResponse }
    }

    private val ImapResponse.isByeResponse: Boolean
        get() = !isTagged && isNotEmpty() && ImapResponseParser.equalsIgnoreCase(first(), Responses.BYE)
}
